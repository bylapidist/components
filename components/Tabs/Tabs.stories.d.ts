import type { StoryObj } from '@storybook/react';
import { BaseProps } from '../types';
import { TabsProps } from './index';
declare const _default: import("@storybook/types").ComponentAnnotations<import("@storybook/react/dist/types-0a347bb9").R, import("@storybook/types").Args>;
export default _default;
export declare const Basic: StoryObj<BaseProps & TabsProps>;
export declare const Small: StoryObj<BaseProps & TabsProps>;
export declare const Medium: StoryObj<BaseProps & TabsProps>;
export declare const Large: StoryObj<BaseProps & TabsProps>;
