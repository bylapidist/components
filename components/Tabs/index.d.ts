import { BaseProps, SizeType } from '../types';
import { LinkProps } from '../Link';
export interface TabsProps {
    readonly size?: SizeType;
    readonly links?: (LinkProps & BaseProps)[];
    readonly onChangeTab?: (activeTab: number) => void;
}
declare const Tabs: {
    ({ className, as, testId, size, links, onChangeTab, ...restProps }: BaseProps & TabsProps): import("react/jsx-runtime").JSX.Element;
    displayName: string;
};
export default Tabs;
