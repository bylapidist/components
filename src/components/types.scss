$levels: 'primary', 'secondary', 'tertiary', 'danger';
$sizes: 'none', 'small', 'medium', 'large';
$families: 'body', 'heading', 'serif', 'mono';
$weights: 'regular', 'medium', 'bold';
