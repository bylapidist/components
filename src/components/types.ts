import {
    ElementType,
    ReactNode,
    CSSProperties,
    ComponentPropsWithoutRef
} from 'react';

export type CommonComponentProps<T extends ElementType> = {
    as?: T;
    testId?: string;
    children?: ReactNode;
    className?: string;
    style?: CSSProperties;
};

export type CommonComponentCombinedProps<
    T extends ElementType,
    K
> = CommonComponentProps<T> &
    Omit<ComponentPropsWithoutRef<T>, keyof CommonComponentProps<T>> &
    K;

export type LevelType = 'primary' | 'secondary' | 'tertiary' | 'danger';

export type SizeType = 'none' | 'small' | 'medium' | 'large';
