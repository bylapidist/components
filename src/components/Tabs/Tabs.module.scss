@import '../global.module';

.this {
    display: flex;
}

.inner {
    display: flex;
    flex-direction: row;
    grid-gap: var(--gutter-small);
    background-color: var(--colour-core-lightest);
}

.active {
    background-color: var(--colour-neutral-lightest);
    box-shadow: var(--elevation-lowest);
}

.link {
    border: transparent solid var(--border-width-thinner);
    border-radius: var(--border-radius-squarer);
}

.link:focus {
    border: var(--colour-secondary-dark) solid var(--border-width-thinner);
}

