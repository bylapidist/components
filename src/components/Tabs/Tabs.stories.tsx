import React from 'react';
import type { Meta, StoryObj, StoryFn } from '@storybook/react';
import { BaseProps } from '../types';
import ThemeProvider from '../ThemeProvider';
import Tabs, { TabsProps } from './index';

export default {
    title: 'Primitives/Tabs',
    component: Tabs,
    decorators: [(getStory) => <ThemeProvider>{getStory()}</ThemeProvider>]
} as Meta;

const defaultArgs: BaseProps & TabsProps = {
    as: 'div',
    testId: 'Tabs',
    size: 'medium',
    links: [
        { testId: 'Link1', href: 'javascript:void', children: 'Home' },
        { testId: 'Link2', href: 'javascript:void', children: 'About' },
        { testId: 'Link3', href: 'javascript:void', children: 'Blog' },
        { testId: 'Link4', href: 'javascript:void', children: 'Contact' }
    ],
    onChangeTab: () => {}
};

const DefaultTemplate: StoryFn<BaseProps & TabsProps> = (args) => (
    <Tabs {...args} />
);

export const Basic: StoryObj<BaseProps & TabsProps> = DefaultTemplate.bind({});
Basic.args = {
    ...defaultArgs,
    onChangeTab: undefined
};

export const Small: StoryObj<BaseProps & TabsProps> = DefaultTemplate.bind({});
Small.args = {
    ...defaultArgs,
    size: 'small'
};
export const Medium: StoryObj<BaseProps & TabsProps> = DefaultTemplate.bind({});
Medium.args = {
    ...defaultArgs,
    size: 'medium'
};
export const Large: StoryObj<BaseProps & TabsProps> = DefaultTemplate.bind({});
Large.args = {
    ...defaultArgs,
    size: 'large'
};
