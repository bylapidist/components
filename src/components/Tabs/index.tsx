import React from 'react';
import cx from 'classnames';
import type { BaseProps, SizeType } from '../types';
import Box from '../Box';
import Text from '../Text';
import Link, { LinkProps } from '../Link';
import styles from './Tabs.module.scss';

export interface TabsProps {
    readonly size?: SizeType;
    readonly links?: (LinkProps & BaseProps)[];
    readonly onChangeTab?: (activeTab: number) => void;
}

const Tabs = ({
    className,
    as = 'div',
    testId = 'Tabs',
    size = 'medium',
    links = [],
    onChangeTab,
    ...restProps
}: BaseProps & TabsProps) => {
    const [activeTab, setActiveTab] = React.useState<number | undefined>(
        undefined
    );

    const handleSetActiveTab = (selectedIndex: number) => {
        if (typeof onChangeTab === 'function') {
            setActiveTab(selectedIndex);
            onChangeTab(selectedIndex);
        }
    };

    return (
        <Text
            as={as}
            data-testid={testId}
            size={size}
            weight="medium"
            className={cx(styles.this, className)}
            {...restProps}
        >
            <Box
                as="ul"
                gutter="medium"
                elevation="large"
                roundness="small"
                data-testid={`${testId}-list`}
                className={cx(styles.inner)}
            >
                {links.map((linkProps, index) => (
                    <Box
                        data-testid={`${testId}-item`}
                        as="li"
                        roundness="small"
                        key={index}
                        className={cx({
                            [styles.active]: index === activeTab
                        })}
                    >
                        <Link
                            data-testid={`${testId}-link`}
                            size={size}
                            onClick={() => handleSetActiveTab(index)}
                            className={cx(styles.link)}
                            {...linkProps}
                        >
                            <Box gutter="medium">{linkProps.children}</Box>
                        </Link>
                    </Box>
                ))}
            </Box>
        </Text>
    );
};

Tabs.displayName = 'Tabs';
export default Tabs;
