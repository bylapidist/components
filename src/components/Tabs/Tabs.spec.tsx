import React from 'react';
import { vi, describe, expect, test, afterEach } from 'vitest';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import { SizeType } from '../types';
import Tabs from './index';

const setup = (Component: React.ReactElement) => render(Component);

const mockLinks = [
    { children: 'Tab1', href: '#tab1' },
    { children: 'Tab2', href: '#tab2' }
];

afterEach(cleanup);

const sizes: SizeType[] = ['small', 'medium', 'large'];
describe('Tabs', () => {
    test('it works with defaults', () => {
        const { container } = setup(<Tabs links={mockLinks} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test.each(sizes)('it works with sizes', (size) => {
        const { container } = setup(<Tabs links={mockLinks} size={size} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('it works with default testId', () => {
        setup(<Tabs links={mockLinks} />);
        expect(screen.getByTestId('Tabs')).toBeTruthy();
    });

    test('it works with specified testId', () => {
        setup(<Tabs links={mockLinks} testId="TestId" />);
        expect(screen.getByTestId('TestId')).toBeTruthy();
    });

    test('it works with className', () => {
        setup(<Tabs links={mockLinks} className="test" />);
        expect(
            screen.getByTestId('Tabs').classList.contains('test')
        ).toBeTruthy();
    });

    test('updates activeTab state and calls onChangeTab when a tab is clicked', () => {
        const handleChangeTab = vi.fn();
        setup(<Tabs links={mockLinks} onChangeTab={handleChangeTab} />);

        fireEvent.click(screen.getByText('Tab2'));
        expect(handleChangeTab).toHaveBeenCalledWith(1);
    });
});
