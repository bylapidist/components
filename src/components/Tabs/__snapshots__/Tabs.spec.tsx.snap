// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > it works with defaults 1`] = `
<div
  class="_this_b26e6e _size-medium_b26e6e _family-body_b26e6e _weight-medium_b26e6e _this_40c7a2"
  data-testid="Tabs"
>
  <ul
    class="_this_d62db4 _elevation-large_d62db4 _roundness-small_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_40c7a2"
    data-testid="Tabs-list"
  >
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-medium_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab1"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab1
          </div>
        </span>
      </a>
    </li>
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-medium_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab2"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab2
          </div>
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Tabs > it works with sizes 1`] = `
<div
  class="_this_b26e6e _size-small_b26e6e _family-body_b26e6e _weight-medium_b26e6e _this_40c7a2"
  data-testid="Tabs"
>
  <ul
    class="_this_d62db4 _elevation-large_d62db4 _roundness-small_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_40c7a2"
    data-testid="Tabs-list"
  >
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-small_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab1"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab1
          </div>
        </span>
      </a>
    </li>
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-small_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab2"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab2
          </div>
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Tabs > it works with sizes 2`] = `
<div
  class="_this_b26e6e _size-medium_b26e6e _family-body_b26e6e _weight-medium_b26e6e _this_40c7a2"
  data-testid="Tabs"
>
  <ul
    class="_this_d62db4 _elevation-large_d62db4 _roundness-small_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_40c7a2"
    data-testid="Tabs-list"
  >
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-medium_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab1"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab1
          </div>
        </span>
      </a>
    </li>
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-medium_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab2"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab2
          </div>
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Tabs > it works with sizes 3`] = `
<div
  class="_this_b26e6e _size-large_b26e6e _family-body_b26e6e _weight-medium_b26e6e _this_40c7a2"
  data-testid="Tabs"
>
  <ul
    class="_this_d62db4 _elevation-large_d62db4 _roundness-small_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_40c7a2"
    data-testid="Tabs-list"
  >
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-large_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab1"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab1
          </div>
        </span>
      </a>
    </li>
    <li
      class="_this_d62db4 _elevation-none_d62db4 _roundness-small_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
      data-testid="Tabs-item"
    >
      <a
        class="_this_b26e6e _size-large_b26e6e _family-body_b26e6e _weight-regular_b26e6e _this_a14c64 _link_40c7a2"
        data-testid="Tabs-link"
        href="#tab2"
      >
        <span
          class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-none_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4 _inner_a14c64"
          data-testid="Link-inner"
        >
          <div
            class="_this_d62db4 _elevation-none_d62db4 _roundness-none_d62db4 _gutter-medium_d62db4 _gutterX-none_d62db4 _gutterY-none_d62db4"
            data-testid="Box"
          >
            Tab2
          </div>
        </span>
      </a>
    </li>
  </ul>
</div>
`;
